File: python/tinyssb/goset.py
import hashlib
from collections import deque
from . import node, io, repo, util #NODE, LOGTYPE_remote

import time

import functools
import os

from typing import Any, Callable, Optional
import bipf


GOSET_DMX_STR = "tinySSB-0.1 GOset 1"
DMX_PFX = b"tinyssb-v0"


class Claim:
    typ = b'c'
    lo = bytes(0)  # util.FID_LEN
    hi = bytes(0)  # util.FID_LEN
    xo = bytes(0)  # util.FID_LEN
    sz = 0
    wire = bytes(0)


class Novelty:
    typ = b'n'
    key = bytes(util.FID_LEN)
    wire = bytes(0)


NOVELTY_LEN = 33 # sizeof(struct novelty_s)
CLAIM_LEN = 98 # sizeof(struct claim_s)
ZAP_LEN = 33 # sizeof(struct zap_s)

GOSET_KEY_LEN   = util.FID_LEN
GOSET_MAX_KEYS  =    100
GOSET_ROUND_LEN = 10
MAX_PENDING     =     20
NOVELTY_PER_ROUND =    1
ASK_PER_ROUND   =      1
HELP_PER_ROUND  =      2
ZAP_ROUND_LEN   =   4500






class GOset():

    def __init__(self, node, dmx_str: str =GOSET_DMX_STR, epoch: int = 0,
                 add_key_callback: Optional[Callable[[bytes], None]] = None) -> None:
        self.keys = []
        self.state = bytearray(util.FID_LEN)
        self.pending_claims = []
        self.pending_novelty = deque()
        self.largest_claim_span = 0
        self.novelty_credit = 1

        self.node = node
        self.dmx_str = dmx_str # for isp: dmx_str = GOSET_DMX_STR + contractID
        self.epoch = epoch
        self.is_root_goset = dmx_str == GOSET_DMX_STR
        self.goset_dmx = hashlib.sha256(self.dmx_str.encode() + (epoch.to_bytes() if not self.is_root_goset else b"") ).digest()[:util.DMX_LEN]
        print("dmx str:", self.goset_dmx.hex())
        self.want_dmx = self.compute_dmx('want'.encode('utf-8') + self.state)
        self.chunk_dmx = self.compute_dmx('blob'.encode('utf-8') + self.state)
        # self.novelt_and_claim_dmx = hashlib.sha256(self.dmx_str.encode()).digest()[:util.DMX_LEN]
        self.node.arm_dmx(self.goset_dmx, lambda pkt, aux: self.rx(pkt))
        self.add_key_callback = add_key_callback
        

    # state = bytearray(util.FID_LEN)
    # pending_claims = []
    # pending_novelty = deque()
    # largest_claim_span = 0
    # novelty_credit = 1

    # def loop(self) -> None:
    #     while True:
    #         self.beacon()
    #         time.sleep(GOSET_ROUND_LEN)


    def rx(self, pkt: bytes, aux: Optional[bytes] = None) -> None:
        print("Goset received", len(pkt))
        
        if len(pkt) <= util.DMX_LEN:
            return
        
        dmx = pkt[:util.DMX_LEN]

        print("received len:", len(pkt))
        print("dmx:", dmx.hex(), "but expected:", self.goset_dmx.hex())

        # sanity check, GosetHandler should ensure this
        if dmx != self.goset_dmx: 
            print("Unknown novelty / claim dmx")
            return
        
        pkt = pkt[util.DMX_LEN:] # remove dmx

        if len(pkt) == NOVELTY_LEN and pkt[0] == ord('n'):

            print("received Novelty")
            self._add_key(pkt[1:NOVELTY_LEN])
            return
        
        if len(pkt) != CLAIM_LEN or pkt[0] != ord('c'):
            print("not claim")
            return
        
        cl = self._mk_claim_from_bytes(pkt)

        if cl.sz > self.largest_claim_span:
            self.largest_claim_span = cl.sz
        if cl.sz == len(self.keys) and util.byteArrayCmp(self.state, cl.xo) == 0:
            print("GOset rx(): seems we are synced (with at least someone), |GOset|=",
                   len(self.keys))
        else:
            self._add_key(cl.lo)
            self._add_key(cl.hi)
            self._add_pending_claim(cl)
            print("add pending claim")


    def beacon(self) -> None:
        print("Beacon called", self.dmx_str)
        if len(self.keys) == 0:
            return
        while(self.novelty_credit > 0 and len(self.pending_novelty) > 0):
            self.novelty_credit -= 1
            self._enqueue(self.pending_novelty.popleft().wire, self.goset_dmx, None)
        self.novelty_credit = NOVELTY_PER_ROUND
        cl = self._mk_claim(0, len(self.keys) - 1)
        if util.byteArrayCmp(cl.xo, self.state) != 0:
            print("GOset state change to", cl.xo.hex(), "|keys|=", len(self.keys))
            self.state = cl.xo
            self.update_want_dmx()
        self._enqueue(cl.wire, self.goset_dmx, None)

        self.pending_claims.sort(key=lambda x: x.sz)
        max_ask = ASK_PER_ROUND
        max_help = HELP_PER_ROUND

        retain: list[Claim] = []
        for c in self.pending_claims:
            if c.sz == 0:
                return
            lo = next(i for i, x in enumerate(self.keys) if util.byteArrayCmp(x, c.lo) == 0)
            hi = next(i for i, x in enumerate(self.keys) if util.byteArrayCmp(x, c.hi) == 0)
            if lo == -1 or hi == -1 or lo > hi:
                continue
            partial = self._mk_claim(lo,hi)
            if util.byteArrayCmp(partial.xo, c.xo) == 0:
                continue
            if partial.sz <= c.sz:
                if max_ask > 0:
                    self._enqueue(partial.wire, self.goset_dmx, None)
                    max_ask -= 1
                if partial.sz < c.sz:
                    retain.append(c)
                    continue

            if max_help > 0:
                max_help -= 1
                hi -= 1
                lo += 1
                if hi <= lo:
                    self._enqueue(self._mk_novelty_from_key(self.keys[lo]).wire, 
                                  self.goset_dmx, None)
                elif hi - lo <= 2:
                    self._enqueue(self._mk_claim(lo,hi).wire, self.goset_dmx, None)
                else:
                    sz = (hi + 1 -lo) // 2
                    self._enqueue(self._mk_claim(lo,lo+sz-1).wire, self.goset_dmx, None)
                    self._enqueue(self._mk_claim(lo+sz, hi).wire, self.goset_dmx, None)
                continue
            retain.append(c)

        while len(retain) >= MAX_PENDING - 5:
            retain.pop()
        self.pending_claims = retain


    def _include_key(self, key: bytes) -> bool:
        zero = bytes(GOSET_KEY_LEN)
        if key == zero:
            return False
        if key in self.keys:
            print("GOset _include_key(): key already exists")
            return False
        if len(self.keys) >= GOSET_MAX_KEYS:
            print("GOset _include_key(): too many keys")
            return False
        print("GOset _include_key", key)
        self.keys.append(key)
        if self.add_key_callback is not None:
            print("call callback")
            self.add_key_callback(key)
        return True


    def _add_key(self, key: bytes) -> None:
        if not self._include_key(key):
            return
        
        typ = repo.FEED_TYPE_ROOT if self.is_root_goset else repo.FEED_TYPE_ISP_VIRTUAL
        if not self.node.repo.feed_exists(key):
            self.node.repo.new_feed(key, typ)

        self.keys = sorted(self.keys, key=functools.cmp_to_key(util.byteArrayCmp))
        print("ADDKEY: ", self.keys)
        if len(self.keys) >= self.largest_claim_span:
            n = self._mk_novelty_from_key(key)
            if self.novelty_credit > 0:
                self._enqueue(n.wire, self.goset_dmx)
                self.novelty_credit -= 1
            elif len(self.pending_novelty) < MAX_PENDING:
                self.pending_novelty.append(n)
        print("GOSET _add_key(): added key", key)


    def _add_pending_claim(self, cl: Claim) -> None:
        for c in self.pending_claims:
            if c.sz == cl.sz and c.xo == cl.xo:
                return
        self.pending_claims.append(cl)


    def _mk_novelty_from_key(self, key: bytes) -> Novelty:
        n = Novelty()
        n.wire = bytes(b'n') + key
        n.key = key
        return n


    def _mk_claim_from_bytes(self, pkt: bytes) -> Claim:
        cl = Claim()
        cl.lo = pkt[1:33]
        cl.hi = pkt[33:65]
        cl.xo = pkt[65:97]
        cl.sz = pkt[97]
        cl.wire = pkt
        return cl


    def _mk_claim(self, lo: int, hi: int) -> Claim:
        cl = Claim()
        cl.lo = self.keys[lo]
        cl.hi = self.keys[hi]
        cl.xo = self._xor(lo, hi)
        cl.sz = hi - lo + 1
        b = cl.sz.to_bytes()
        cl.wire = cl.typ + cl.lo + cl.hi + cl.xo + b
        return cl


    def _xor(self, lo: int, hi: int) -> bytearray:
        xor = bytearray(self.keys[lo])
        for k in self.keys[lo+1 : hi+1]:
            for i in range(len(xor)):
                xor[i] ^= k[i]
        return xor


    def _enqueue(self, buf: bytes, dmx: Optional[bytes] = None,
                  aux: Optional[bytes] = None) -> None:
        pkt = buf if dmx is None else dmx + buf
        for f in self.node.faces:
            f.enqueue(pkt)


    def adjust_state(self) -> None:
        self.keys = sorted(self.keys, key=functools.cmp_to_key(util.byteArrayCmp))
        if len(self.keys) > 0:
            cl = self._mk_claim(0, len(self.keys)-1)
            self.state = cl.xo
        else:
            self.state = bytes(util.FID_LEN)
        print("GOset adjust_state() for", len(self.keys), "resulted in", self.state.hex())
        self.update_want_dmx()

    @staticmethod
    def compute_dmx(buf: bytes) -> bytes:
        return hashlib.sha256(DMX_PFX + buf).digest()[:util.DMX_LEN]
    
    def set_add_key_callback(self, callback: Callable[[bytes], None]) -> None:
        self.add_key_callback = callback
        print("updated callback")

    def set_epoch(self, new_epoch: int):
        if new_epoch < self.epoch:
            raise Exception("GOset epoch can't be decreased")
        
        self.node.arm_dmx(self.goset_dmx, None, None) # remove old goset dmx handler
        self.epoch = new_epoch
        self.goset_dmx = hashlib.sha256(self.dmx_str.encode() + (self.epoch.to_bytes() if not self.is_root_goset else b"") ).digest()[:util.DMX_LEN]
        self.node.arm_dmx(self.goset_dmx, lambda pkt, aux: self.rx(pkt))

    def update_want_dmx(self) -> None:
        print("update want")
        self.node.arm_dmx(self.want_dmx, None, None)
        self.node.arm_dmx(self.chunk_dmx, None, None)

        self.want_dmx = self.compute_dmx('want'.encode('utf-8') + self.state)
        self.chunk_dmx = self.compute_dmx('blob'.encode('utf-8') + self.state)

        print("set want to:", self.want_dmx.hex())

        self.node.arm_dmx(self.want_dmx,
                          lambda buf, aux: self.incoming_want_request(buf, None, None))
        self.node.arm_dmx(self.chunk_dmx,
                          lambda buf, aux: self.incoming_chunk_request(buf, None, None))
        
    def incoming_want_request(self, buf: bytes, aux:Optional[bytes] = None, neigh: Optional[io.FACE] = None) -> None:
        """
        Handle want request.
        
        """
        print("incoming want")

        lst = bipf.loads(buf[util.DMX_LEN:])
        if not lst or type(lst) is not list:
            print("error decoding want request")
            return
        if len(lst) < 1 or type(lst[0]) is not int:
            print("error decoding want request with offset")
            return
        offs = lst[0]
        v = "WANT vector=["
        credit = 3
        for i in range(1, len(lst)):
            try:
                ndx = (offs + i - 1) % len(self.keys)
                fid = self.keys[ndx]
                seq = lst[i]
                v += f' {ndx}.{seq}'
            except:
                print("error incoming Want error")
                continue
            while credit > 0:
                pkt = self.node.repo.feed_read_pkt(fid, seq)
                if pkt is None:
                    break
                print("NODE have entry", fid.hex(),".",seq)
                for f in self.node.faces:
                    f.enqueue(pkt)
                seq += 1
                credit -= 1

        v += " ]"
        print("Node", v)
        if credit == 3:
            print("Node no entry found to serve")

    def incoming_chunk_request(self, buf: bytes, aux: Optional[bytes], neigh: Optional[io.FACE]) -> None:
        print("Node incoming CHNK request")
        vect = bipf.loads(buf[util.DMX_LEN:])
        if vect == None or type(vect) != list:
            print("Node error decoding CHNK request")
            return
        print(vect)
        v = "CHNK vector=["
        credit = 3
        for e in vect:
            try:
                fNDX = e[0]
                fid = self.keys[fNDX]
                seq = e[1]
                cnr = e[2]
                v += f" {fid.hex()}.{seq}.{cnr}"
            except Exception as e:
                print("Node incoming CHNK error")
                print(e)
                continue
            
            pkt = self.node.repo.feed_read_pkt(fid, seq)
            if pkt is None or pkt[util.DMX_LEN] != util.PKTTYPE_chain20:
                continue
            (sz, szlen) = bipf.varint_decode_max(pkt, util.DMX_LEN + 1, util.DMX_LEN + 4)
            if(sz <= 28 - szlen):
                continue
            maxChunks = (sz - (28 - szlen) + 99) // 100
            if cnr > maxChunks:
                continue
            while(cnr <= maxChunks and credit > 0):
                credit -= 1
                chunk = self.node.repo.feed_read_chunk(fid, seq, cnr)
                print("CHUNK:", chunk)
                if chunk is None:
                    print("could not find chunk")
                    break
                for f in self.node.faces:
                    f.enqueue(chunk)
                cnr += 1
        v += " ]"
        print("Node", v)


class GOsetManager():
    def __init__(self, node) -> None:
        self.sets: set[GOset] = set()
        self.node = node
        self.offs: dict[GOset, int] = {}

    def loop(self) -> None:
        while True:
            print("amount of gosets:", len(self.sets))
            for go in self.sets:
                print("Beacon for", go.dmx_str)
                go.beacon()
            time.sleep(GOSET_ROUND_LEN)

    def add_goset(self, dmx_str: str =GOSET_DMX_STR, epoch: int = 0,
                 add_key_callback: Optional[Callable[[bytes], None]] = None) -> GOset:
        go = GOset(self.node, dmx_str, epoch, add_key_callback)
        self.sets.add(go)
        self.offs[go] = 0

        return go

    def remove_goset(self, goset: GOset) -> bool:
        for go in self.sets:
            if goset == go:
                self.node.arm_dmx(go.goset_dmx) # rmv goset_dmx handler
                self.sets.remove(go)
                del self.offs[go]
                return True
        return False

    

File: android/app/src/main/java/nz/scuttlebutt/tremolavossbol/tssb/GOset.kt
package nz.scuttlebutt.tremolavossbol.tssb

import android.util.Log
import nz.scuttlebutt.tremolavossbol.MainActivity
import nz.scuttlebutt.tremolavossbol.crypto.SodiumAPI.Companion.sha256
import nz.scuttlebutt.tremolavossbol.utils.Bipf
import nz.scuttlebutt.tremolavossbol.utils.Constants
import nz.scuttlebutt.tremolavossbol.utils.Constants.Companion.DMX_LEN
import nz.scuttlebutt.tremolavossbol.utils.Constants.Companion.FID_LEN
import nz.scuttlebutt.tremolavossbol.utils.Constants.Companion.GOSET_DMX_STR
import nz.scuttlebutt.tremolavossbol.utils.HelperFunctions.Companion.byteArrayCmp
import nz.scuttlebutt.tremolavossbol.utils.HelperFunctions.Companion.fromEncodedUByte
import nz.scuttlebutt.tremolavossbol.utils.HelperFunctions.Companion.toEncodedUByte
import nz.scuttlebutt.tremolavossbol.utils.HelperFunctions.Companion.toHex
import kotlin.experimental.xor

class Claim {
    var typ: Byte = 'c'.toByte()
    var lo = ByteArray(0) // FID_LEN)
    var hi = ByteArray(0) // FID_LEN)
    var xo = ByteArray(0) // FID_LEN)
    var sz = 0
    var wire = ByteArray(0)
}

class Novelty {
    var typ: Byte = 'n'.toByte()
    var key = ByteArray(FID_LEN)
    var wire = ByteArray(0)
}

class GOset(val context: MainActivity, val dmx_str: String = GOSET_DMX_STR, val epoch: Int = 0) {
    /* packet format:
        n 32B 32B? 32B?  // 33 bytes, in the future up to two additional keys
        c 32B 32B 32B B  // 98 bytes
          lo  hi  xor cnt(4 bytes, big endian)
        z 32B(nonce)     // 33 bytes
    */

    val NOVELTY_LEN = 33 // sizeof(struct novelty_s)
    val CLAIM_LEN = 98   // sizeof(struct claim_s)
    val ZAP_LEN = 33     // sizeof(struct zap_s)

    val GOSET_KEY_LEN   = FID_LEN
    val GOSET_MAX_KEYS  =    100 //   MAX_FEEDS
    val GOSET_ROUND_LEN = 10000L // in millis
    val MAX_PENDING     =     20 // log2(maxSetSize) + density (#neighbors)
    val NOVELTY_PER_ROUND =    1
    val ASK_PER_ROUND   =      1
    val HELP_PER_ROUND  =      2
    val ZAP_ROUND_LEN   =   4500

    var state = ByteArray(FID_LEN)

    val str = dmx_str + if (epoch == 0) "" else epoch
    var goset_dmx = str.encodeToByteArray().sha256().sliceArray(0..DMX_LEN-1)
    var want_dmx = context.tinyDemux.compute_dmx("want".encodeToByteArray() + state)
    var chunk_dmx = context.tinyDemux.compute_dmx("blob".encodeToByteArray() + state)
    val is_root_goset = dmx_str == GOSET_DMX_STR
    // unsigned short version; ??
    // const char* fname;
    // int goset_len; // number of set elements

    val keys = ArrayList<ByteArray>(0)
    var pending_claims = ArrayList<Claim>(0)
    val pending_novelty = ArrayDeque<Novelty>(0)
    var largest_claim_span = 0
    var novelty_credit = 1
    // unsigned long next_round; // FIXME: handle, or test, wraparound
    // struct zap_s zap;
    // unsigned long zap_state;
    // unsigned long zap_next;

    fun rx(pkt: ByteArray, aux: ByteArray?) {
        Log.d("goset", "rx ${pkt.size}")

        if (pkt.size <= DMX_LEN)
            return
        val buf = pkt.sliceArray(DMX_LEN .. pkt.lastIndex)

        if (buf.size == NOVELTY_LEN && buf[0] == 'n'.toByte()) {
            _add_key(buf.sliceArray(1..NOVELTY_LEN-1))
            return
        }
        /*
        if (pkt[0] == 'z' && len == ZAP_LEN) {
            Serial.println("ZAP message received");
            unsigned long now = millis();
            if (gp->zap_state == 0) {
                Serial.println("ZAP phase I starts");
                memcpy(&gp->zap, pkt, ZAP_LEN);
                gp->zap_state = now + ZAP_ROUND_LEN;
                gp->zap_next = now;
            }
            return;
        }
        */
        if (buf.size != CLAIM_LEN || buf[0] != 'c'.toByte())
            return;
        val cl = mkClaim(buf)
        if (cl.sz > largest_claim_span)
            largest_claim_span = cl.sz
        // Log.d("rx", "state = ${state.toHex()}, cl.sz=${cl.sz}, k.sz=${keys.size}")
        if (cl.sz == keys.size && byteArrayCmp(state, cl.xo) == 0) {
            Log.d("goset", "seems we are synced (with at least someone), |GOset|=${keys.size}")
        } else {
            _add_key(cl.lo)
            _add_key(cl.hi)
            _add_pending_claim(cl)
        }
        // Log.d("goset", "goset size=${keys.size}, pending_claims.size=${pending_claims.size}")
    }

    fun loop() {
        while (true) {
            beacon()
            Thread.sleep(GOSET_ROUND_LEN)
        }
    }

    fun beacon() { // called in regular intervals
        /*
        unsigned long now = millis();
        if (gp->zap_state != 0) {
            if (now > gp->zap_state + ZAP_ROUND_LEN) { // two rounds after zap started
            Serial.println("ZAP phase II ended, resetting now");
            repo_reset();
        }
            if (now < gp->zap_state && now > gp->zap_next) { // phase I
            Serial.println("sending zap message");
            io_send(_mkZap(gp), DMX_LEN + ZAP_LEN, NULL);
            gp->zap_next = now + 1000;
        }
        }
        */
        // Log.d("GOset", "beacon")
        Log.d("goset", "goset_dmx: ${goset_dmx.toHex()}")
        Log.d("goset", "goset_string: $str")
        if (keys.size == 0) return
        while (novelty_credit-- > 0 && pending_novelty.size > 0)
            context.tinyIO.enqueue(pending_novelty.removeFirst().wire, goset_dmx, null)
        novelty_credit = NOVELTY_PER_ROUND

        val cl = mkClaim(0, keys.size-1)
        if (byteArrayCmp(cl.xo, state) != 0) { // GOset changed
            Log.d("goset", "state change to ${cl.xo.toHex()}, |keys|=${keys.size}")
            state = cl.xo
            context.tinyDemux.set_want_dmx(this)
        }
        context.tinyIO.enqueue(cl.wire, goset_dmx, null);

        // sort pending entries, smallest first
        pending_claims.sortBy({c -> c.sz})
        var max_ask = ASK_PER_ROUND
        var max_help = HELP_PER_ROUND

        val retain = ArrayList<Claim>()
        for (c in pending_claims) {
            Log.d("goset", "pending claim loop")
            if (c.sz == 0) continue // ignore bogous claims
            var lo = keys.indexOfFirst({k -> byteArrayCmp(k,c.lo) == 0})
            var hi = keys.indexOfFirst({k -> byteArrayCmp(k,c.hi) == 0})
            if (lo == -1 || hi == -1 || lo > hi) continue // ignore bogous claims
            val partial = mkClaim(lo, hi)
            if (byteArrayCmp(partial.xo, c.xo) == 0) continue // match, hence no reason to ask others
            if (partial.sz <= c.sz) { // ask for help, starting with the smallest entry, for a limited number of questions
                if (max_ask-- > 0) {
                    // Serial.print("asking for help " + String(partial->cnt));
                    // Serial.print(String(" ") + to_hex(partial->lo,4) + String(".."));
                    // Serial.println(String(" ") + to_hex(partial->hi,4) + String(".."));
                    context.tinyIO.enqueue(partial.wire, goset_dmx, null);
                }
                if (partial.sz < c.sz) {
                    retain.add(c)
                    continue;
                }
            }
            if (max_help-- > 0) { // we have larger claim span, offer help (but limit # of claims)
                hi--; lo++
                // Log.d("goset","offer help span= " + String(partial->cnt - 2));
                // Serial.print(String(" ") + to_hex(gp->goset_keys+lo*GOSET_KEY_LEN,4) + String(".."));
                // Serial.println(String(" ") + to_hex(gp->goset_keys+hi*GOSET_KEY_LEN,4) + String(".."));
                Log.d("goset", "max_help ${lo}, ${hi}")
                if (hi <= lo)
                    context.tinyIO.enqueue(mkNovelty_from_key(keys[lo]).wire, goset_dmx, null)
                else if (hi - lo <= 2) { // span of 2 or 3
                    context.tinyIO.enqueue(mkClaim(lo, hi).wire, goset_dmx, null)
                }
                else { // split span in two intervals
                    val sz = (hi+1 - lo) / 2
                    context.tinyIO.enqueue(mkClaim(lo, lo+sz-1).wire, goset_dmx, null)
                    context.tinyIO.enqueue(mkClaim(lo+sz, hi).wire, goset_dmx, null)
                }
                continue; // only help once per claim, do not retain
            }
            retain.add(c) // we could not treat the claim in this round, hence retain
        }
        // make room for new claims
        while (retain.size >= MAX_PENDING-5)
            retain.removeLast()
        pending_claims = retain
        // Serial.println(String(gp->pending_c_cnt,DEC) + " pending claims, GOset size is " + String(gp->goset_len, DEC));
        // for (int i = 0; i < gp->pending_c_cnt; i++)
        // Serial.println("  pend claim span=" + String(gp->pending_claims[i].cnt, DEC)
        // + " " + to_hex(gp->pending_claims[i].xo,32));
    }

    fun _include_key(key: ByteArray): Boolean {
        val zero = ByteArray(GOSET_KEY_LEN)
        if (key.contentEquals(zero))
            return false
        for (k in keys)
            if (k.contentEquals(key)) {
                // Log.d("goset", "key ${key.toHex()} already exists")
                return false
            }
        if (keys.size >= GOSET_MAX_KEYS) {
            Log.d("goset", "too many keys: ${keys.size}")
            return false
        }
        Log.d("goset","_include_key ${key.toHex()}, keys.size was ${keys.size}")
        keys.add(key)
        return true
    }

    fun _add_key(key: ByteArray) {
        if (!_include_key(key)) // adds key if necessary
            return
        val type = if(is_root_goset) context.tinyRepo.FEED_TYPE_ROOT else context.tinyRepo.FEED_TYPE_ISP_VIRTUAL

        if (!context.tinyRepo.feed_exists(key))
            context.tinyRepo.new_feed(key, type)

        keys.sortWith({a:ByteArray,b:ByteArray -> byteArrayCmp(a,b)})
        if (keys.size >= largest_claim_span) { // only rebroadcast if we are up to date
            val n = mkNovelty_from_key(key)
            if (novelty_credit-- > 0)
                context.tinyIO.enqueue(n.wire, goset_dmx, null)
            else if (pending_novelty.size < MAX_PENDING)
                pending_novelty.add(n)
        }
        context.ble?.refreshShortNameForKey(key) // refresh shortname in devices overview
        Log.d("goset", "added key ${key.toHex()}, |keys|=${keys.size}")
    }

    fun _add_pending_claim(cl: Claim) {
        for (c in pending_claims)
            if (c.sz == cl.sz && c.xo.contentEquals(cl.xo))
                return
        pending_claims.add(cl)
    }

    fun adjust_state() {
        keys.sortWith({a:ByteArray,b:ByteArray -> byteArrayCmp(a,b)})
        if (keys.size > 0) {
            val cl = mkClaim(0, keys.size-1)
            state = cl.xo
        } else
            state = ByteArray(FID_LEN)
        Log.d("goset", "adjust state for ${keys.size} keys, resulted in ${state.toHex()}")
        context.tinyDemux.set_want_dmx(this)
    }

    fun mkClaim(pkt: ByteArray): Claim {
        val cl = Claim()
        cl.lo = pkt.sliceArray(1..32)
        cl.hi = pkt.sliceArray(33..64)
        cl.xo = pkt.sliceArray(65..96)
        cl.sz = pkt[97].fromEncodedUByte()
        cl.wire = pkt
        return cl
    }

    fun mkClaim(lo: Int, hi: Int): Claim {
        val cl = Claim()
        cl.lo = keys[lo]
        cl.hi = keys[hi]
        cl.xo = _xor(lo, hi)
        cl.sz = hi - lo + 1
        val b = cl.sz.toEncodedUByte()
        cl.wire = byteArrayOf(cl.typ) + cl.lo + cl.hi + cl.xo + byteArrayOf(b)
        return cl
    }

    fun mkNovelty(pkt: ByteArray): Novelty {
        val n = Novelty()
        n.key = pkt.sliceArray(1..33)
        n.wire = pkt
        return n
    }

    fun mkNovelty_from_key(key: ByteArray): Novelty {
        val n = Novelty()
        n.wire = byteArrayOf('n'.toByte()) + key
        n.key = key
        return n
    }

    fun _xor(lo: Int, hi: Int): ByteArray {
        val xor = keys[lo].clone()
        for (k in keys.subList(lo+1,hi+1)) // argh, in Kotlin the toIndex is EXCLUSIVE, unlike in a range
            for (i in 0..FID_LEN-1)
                xor[i] = xor[i] xor k[i]
        // Log.d("xor", "lo=${keys[lo].toHex()}, cnt=${keys.size}")
        // Log.d("xor", "hi=${keys[hi].toHex()}")
        // Log.d("xor", "xo=${xor.toHex()}")
        return xor
    }

    fun incoming_want_request(buf: ByteArray, aux: ByteArray?, sender: String?) {
        Log.d("node", "incoming WANT ${buf.toHex()} from sender: $sender")
        val vect = Bipf.bipf_loads(buf.sliceArray(DMX_LEN..buf.lastIndex))
        if (vect == null || vect.typ != Bipf.BIPF_LIST) return
        val lst = vect.getList()
        if (lst.size < 1 || lst[0].typ != Bipf.BIPF_INT) {
            Log.d("node", "incoming WANT error with offset")
            return
        }
        val offs = lst[0].getInt()
        var v = "WANT vector=["
        var vector = mutableMapOf<Int, Int>() //send to frontend
        var credit = 3
        for (i in 1..lst.lastIndex) {
            val fid: ByteArray
            var seq: Int
            try {
                val ndx = (offs + i - 1) % keys.size
                fid = keys[ndx]
                seq = lst[i].getInt()
                v += " $ndx.${seq}"
                vector[ndx] = seq
            } catch (e: Exception) {
                Log.d("node", "incoming WANT error ${e.toString()}")
                continue
            }
            // Log.d("node", "want ${fid.toHex()}.${seq}")
            while (credit > 0) {
                val pkt = context.tinyRepo.feed_read_pkt(fid, seq)
                if (pkt == null)
                    break
                Log.d("node", "  have entry ${fid.toHex()}.${seq}, dmx: ${pkt.sliceArray(0 until DMX_LEN).toHex()}")
                context.tinyIO.enqueue(pkt)
                seq++;
                credit--;
            }
        }
        v += " ]"
        Log.d("node", v)
        if(sender != null)
            context.tinyNode.update_progress(vector.toSortedMap().values.toList(), sender)
        if (credit == 3)
            Log.d("node", "  no entry found to serve")
    }

    fun incoming_chunk_request(buf: ByteArray, aux: ByteArray?) {
        Log.d("node", "incoming CHNK request")
        val vect = Bipf.bipf_loads(buf.sliceArray(DMX_LEN..buf.lastIndex))
        if (vect == null || vect.typ != Bipf.BIPF_LIST) {
            Log.d("node", "  malformed?")
            return
        }
        var v= "CHNK vector=["
        var credit = 3
        for (e in vect.getList()) {
            val fNDX: Int
            val fid: ByteArray
            val seq: Int
            var cnr: Int
            try {
                val lst = e.getList()
                fNDX = lst[0].getInt()
                fid = keys[fNDX]
                seq = lst[1].getInt()
                cnr = lst[2].getInt()
                v += " ${fid.sliceArray(0..9).toHex()}.${seq}.${cnr}"
            } catch (e: Exception) {
                Log.d("node", "incoming CHNK error ${e.toString()}")
                continue
            }
            val pkt = context.tinyRepo.feed_read_pkt(fid, seq)
            if (pkt == null || pkt[DMX_LEN].toInt() != Constants.PKTTYPE_chain20) continue;
            val (sz, szlen) = Bipf.varint_decode(pkt, DMX_LEN + 1, DMX_LEN + 4)
            if (sz <= 28 - szlen) continue;
            val maxChunks    = (sz - (28 - szlen) + 99) / 100
            Log.d("node", "maxChunks is ${maxChunks}")
            if (cnr > maxChunks) continue
            while (cnr <= maxChunks && credit-- > 0) {
                val chunk = context.tinyRepo.feed_read_chunk(fid, seq, cnr)
                if (chunk == null) break;
                Log.d("node", "  have chunk ${fid.sliceArray(0..19).toHex()}.${seq}.${cnr}")
                context.tinyIO.enqueue(chunk);
                cnr++;
            }
        }
        v += " ]"
        Log.d("node", v)
    }
}

class GoSetManager(val context: MainActivity) {
    val sets = ArrayList<GOset>()
    val offs = mutableMapOf<GOset, Int>()
    val GOSET_ROUND_LEN = 10000L

    fun add_goset(dmx_str: String, epoch: Int): GOset {
        val go = GOset(context, dmx_str, epoch)
        sets.add(go)
        offs[go] = 0

        return go
    }

    fun add_goset(goset: GOset) {
        sets.add(goset)
        offs[goset] = 0
    }

    fun remove_goset(goset: GOset): Boolean {
        val removed = sets.remove(goset)
        if (removed)
            offs.remove(goset)
        return removed
    }

    fun loop() {
        while (true) {
            for (set in sets) {
                set.beacon()
                Thread.sleep(GOSET_ROUND_LEN)
            }
        }
    }

}